### ISSUE TITLE
add Codefix that demotes explicit any annotation to {}

### ISSUE THREAD

**Issue by aluanhaddad** [Microsoft/TypeScript#13243]
I would very much appreciate having a codefix (refactoring) that, given an explicit `any` annotation, offers to change it to a `{}` annotation.

For example, given
```ts
function log(x: any) {
  console.log(x);
}
```
a code fix would be offered to change the declared type of `x` to `{}` resulting in
```ts
function log(x: {}) {
  console.log(x);
}
```

**Comment by aleksey-bykov** [comment]
but `any` is more than `{}` technically speaking, if we take it to absurd level it's not different than asking to replace `any` with `HTMLDivElement`

why `{}`?

**Comment by basarat** [comment]
Perhaps he wants `any` => `all the members that are currently accessed`?

```ts
function log(x: any) {
  console.log(x.bar.baz);
}
```
=> 
```ts
function log(x: {bar: {baz: any}}) {
  console.log(x.bar.baz);
}
```

That said the original suggestion is easily done in most editors with `expand selection` + `{` + `delete` shortcuts e.g. [alm](http://alm.tools/):

![tmp](https://cloud.githubusercontent.com/assets/874898/21584087/9fefb478-d0ee-11e6-8d7b-f8933a78b1c7.gif)

I do this quite commonly myself as well to cleanup existing JS code to discover how the argument is actually used :rose:

**Comment by aleksey-bykov** [comment]
@basarat what is the competitive advantage of atom compared to vscode?

**Comment by basarat** [comment]
@aleksey-bykov `alm` (not atom). Most devs have their own setup to help them work better / easily / faster / work the way they do. I just shared mine as alm. Features are covered here https://basarat.gitbooks.io/alm/content/. Trying it is easy `npm install alm -g` so give it a go ... or not .. be happy :heart: :rose:

**Comment by basarat** [comment]
I forgot I wrote the reasons : https://basarat.gitbooks.io/alm/content/contributing/why.html . Been a while since someone has asked :)

**Comment by xLama** [comment]
I think it would be a good idea for tslint

**Comment by aleksey-bykov** [comment]
aside from you usecase which makes sense, `{}` is generally speaking not assignable from `null` and `undefined`, which are quite common valid options, how would it be like outside of your case?

furthermore `any` is a trivial thing to find either by linter (`ts.SyntaxKind.AnyKeyword`) or full text search

**Comment by aleksey-bykov** [comment]
i gave it a second thought, i might be helpful if such code action looked for all places that call this function and tried to deduce the replacement type out of all possible arguments out there
```
// before
function log(value: any) {
}
log(1);
log('a');
```

```
// after code action
log(value: number | string) {
}
log(1);
log('a');
```

**Comment by aluanhaddad** [comment]
I have renamed this issue to reflect the far more useful refactoring proposed by @aleksey-bykov in his remarks above.

The proposed refactoring would look at how a value implicitly typed as `any` is used, determine a candidate type from these use sites, and annotate this value explicitly with that type.

**Comment by mhegazy** [comment]
Addressed by https://github.com/Microsoft/TypeScript/pull/14786

**Comment by amcasey** [comment]
Can this issue be closed?  It sounds like it's been addressed.

**Comment by mhegazy** [comment]
nope. i need to update https://github.com/Microsoft/TypeScript/pull/14786, get it reviewed and merged.

**Comment by aluanhaddad** [comment]
Still super excited about this feature. Thanks @mhegazy

**Comment by ntwilson** [comment]
I just read [#14786](https://github.com/Microsoft/TypeScript/pull/14786), and am also incredibly excited to start using this feature!  Out of curiosity, any reason not to use the same feature for implicitly inferring function parameters rather than assuming any unannotated inputs are type `any`?   In other words, in addition to the Quick Fix, could the compiler use the inferred type instead of `any` even without the annotation?  I feel as though that kind of static analysis could help TypeScript catch a lot of bugs on old javascript code bases that don't have any annotations.

**Comment by ntwilson** [comment]
Sorry, just noticed [#15114](https://github.com/Microsoft/TypeScript/issues/15114) and all its friends that discuss that very issue.  I'm quite happy to see this Quick Fix solution after reading through many of the problems with function parameter inference!
